src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -845,3 +851,39 @@ export const kValue = {
     toMinus32: -Math.pow(2, -32),
   },
 } as const;
+
+/**
+ * @returns a list of test cases where the inputs are the cartesian product of
+ * |values| with itself and the expected value is which ever value appears
+ * earlier in the input.
+ *
+ * Both FTZ and no flush behaviour is considered to be expected for subnormal
+ * floats.
+ *
+ * An example for min, where min(a, b) is expected to return a if a <= b, otherwise b.
+ * |values| would need to be an ascending sorted list, e.g. [-1.0, x, 100.0], where x is a subnormal value
+ * This function will return these test cases:
+ *   {[-1.0, -1.0], -1.0}
+ *   {[-1.0, -0.0], -1.0}
+ *   {[-0.0, -1.0], -1.0}
+ *   {[-1.0, 10.0], -1.0}
+ *   {[10.0, 11.0], -1.0}
+ *   {[0.0, 0.0], anyOf(x, 0.0)}
+ *   {[0.0, 10.0], anyOf(x, 0.0)}
+ *   {[10.0, 0.0], anyOf(x, 0.0)}
+ *   {[10.0, 10.0], 10.0}
+ **/
+export function generateCartesianPairCases(values: Array<Scalar>): CaseList {
+  const cases: CaseList = [];
+  cartesianPairs(values).forEach(inputs => {
```

Comment:
I know you're trying to be DRY here by reusing the cartesian product.  
But the algorithm is a bit obscure, and has a built-in inefficiency in having to look up the element every time.

Open-coding the cartesian product would be clearer, using the form of [forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) where the index is given to the lambda as a parameter:

```js
values.forEach((e,ei) => {
  values.forEach((f,fi) => {
    let precise_expected = (ei <= fi ? e : f);
    let expected = isSubnormalScalar(precise_expected ? anyOf(precise_expected,f32(0.0)) : precise_expected);
    result.push([e,f, expected ]);
  })
});
```